docs/core/modules.rst

```diff
@@ -0,0 +1,132 @@
+.. core_modules:
+
+Modules
+=======
+
+.. module:: superdesk.core
+
+Loading Modules
+---------------
+
+The :class:`Module <module.Module>` class is used to register modules with the system, and provides attributes
+used to define certain behaviour of the module.
+
+An example of a module's root file::
+
+    # file: my/module/__init__py
+    from superdesk.core.module import Module, SuperdeskAsyncApp
+
+    def init(app: SuperdeskAsyncApp):
+        ...
+
+    module = Module(
+        name="my.module",
+        init=init,
+        frozen=True,
+        priority=100
+    )
+
+Then you would add the following to your :ref:`settings.modules` config::
+
+    # file: settings.py
+    MODULES = ["my.module"]
+
+The system would then load this module when the application starts
+
+Module Loading Sequence
+-----------------------
+
+The :class:`SuperdeskAsyncApp <superdesk.core.app.SuperdeskAsyncApp>` app is in charge of loading modules,
+when the `start` function is run on the app.
+
+Modules are loaded in the following 3 stages.
+
+1. Import python module
+2. Load module local configs
+3. Run module init functions
+
+The app will import ALL modules first, then ALL module configs are loaded, then ALL module init functions are run.
+This way we can ensure that all python modules and configs are loaded **before** initialising the modules themself.
+
+Module Configuration
+--------------------
+
+A module can optionally have a :ref:`config <core_config>` loaded on app startup. This can be achieved by setting config options on the Module
+instance.::
+
+    # file: my/module/__init__py
+    from typing import Optional, Dict, Any
+
+    from superdesk.core.module import Module, SuperdeskAsyncApp
+    from superdesk.core.config import ConfigModel
+
+
+    # Create a new Config class that inherits from `ConfigModel`
+    class ModuleConfig(ConfigModel):
+        # Make sure to provide defaults for all fields
+        # otherwise validation would fail
+        uri: str = "localhost"
+        port: int = 27017
+        options: Optional[Dict[str, Any]] = None
+
+    # Create a module local instance of the class,
+    # to be used at runtime
+    config = ModuleConfig()
+
+
+    def init(app: SuperdeskAsyncApp):
+        # The module's config is loaded BEFORE executing the
+        # modules `init` function
+        if config.uri == "localhost":
+            print("Using a local instance")
+        else:
+            print("Using a remote instance")
+
+        # You can also import configs from other modules
+        # all configs are loaded BEFORE executing module's
+        # init function
```

Comment:
I'm not sure if we should try to expose module config to other modules, that kind of breaks module encapsulation. Not that we should try to prevent it, but would probably not document that part.
Maybe we should have some section in general what to do and what not to do.

Reply:
Why can't one module depend on another? Configs are global anyway, this way it's typed and validated on load

Reply:
I don't consider it a public interface of a module, but that might be just my view

Reply:
One reason why I made it a module local instance, is so other files in that module can import it directly and use, with strict typing.

The Configs are frozen by default, so they cannot be changed once loaded. Would that suffice for now, that we document do's and don'ts